### DISCUSSION TITLE
[3 Novembre 2022] - Pipeline

### DISCUSSION THREAD

**Discussion by notedo** [CS-Swap/Architettura-degli-Elaboratori#23] [Pipeline]
### Traccia Esercizio

Il programma qui di seguito somma i numeri negativi nella sequenza fornita.
Si supponga che tutte le istruzioni impiegate fanno parte del set supportato dalla CPU in figura, ossia non si fa uso di alcuna pseudoistruzione.
Si indichino (ignorando hazard che possano concernere la syscall):
1) le istruzioni tra le quali sono presenti data hazard – indicare i numeri di istruzione N ed M (visibili alla sinistra delle linee di codice in figura) ed il registro coinvolto $xY (nel formato N / M / $xY);
2) le istruzioni tra le quali sono presenti control hazard – indicare i numeri di istruzione N ed M (nel formato N / M);
3) quanti cicli di clock sono necessari ad eseguire il programma tramite forwarding, spiegando il calcolo effettuato;
4) quanti cicli di clock sarebbero necessari ad eseguire il programma senza forwarding, spiegando il calcolo effettuato;
5) quali sono, per ognuna delle cinque fasi, le istruzioni (o le bolle) in pipeline durante il 14° ciclo di clock (con forwarding);
6) Si descriva nel maggior dettaglio possibile il funzionamento dei segnali di controllo ForwardA e ForwardB nella fase EXE del diagramma seguente

<img width="566" alt="img" src="https://github.com/CS-Swap/Architettura-degli-Elaboratori/assets/110790224/908f1465-c121-4210-af62-4fa66a0706a8">

<img width="870" alt="img" src="https://github.com/CS-Swap/Architettura-degli-Elaboratori/assets/110790224/5cc1b4ab-cf32-4d27-aa8e-0cdf62669ccd">

### Soluzione prof:

[AE.2021-22.-.Appello.S1.-.2022-11-03.-.Esercizio.pipeline.-.Scheda.di.calcolo.pdf](https://github.com/user-attachments/files/15518314/AE.2021-22.-.Appello.S1.-.2022-11-03.-.Esercizio.pipeline.-.Scheda.di.calcolo.pdf)

**Comment by alem1105** [comment]
## Domanda 6

I segnali `ForwardA` e `ForwardB` servono a gestire i forwarding della pipeline nella fase di **EXE**.
Infatti la **Forwarding EX unit** controllera' per prima cosa due casi:
- Se `ID / EX.rs == EX / MEM.rd` oppure `ID / EX.rt == EX / MEM.rd` attiverà un forwarding sull’istruzione precedente
- Se `ID / EX.rs == MEM / WB.rd` oppure `ID / EX.rt == MEM / WB.rd` attiverà un forwarding sulla seconda istruzione precedente

Nello specifico `ForwardA` si riferisce al registro `rs` mentre `ForwardB` al registro `rt`, quindi nel caso in cui non c'e' forwarding i segnali vengono impostati a `00` e i dati provengono dal blocco `ID / EXE` normalmente.
Se c'e' un forwarding e prendiamo `rs` / `rt` dal blocco `EXE / MEM` allora, rispettivamente, `ForwardA` e / o `ForwardB` verranno impostati a 10.
Nell'ultimo caso, dove prendiamo i registri dal blocco `MEM / WB` impostiamo i segnali a 01.

**Comment by notedo** [comment]
![ESERCIZI - Architettura-84](https://github.com/CS-Swap/Architettura-degli-Elaboratori/assets/110790224/ca6008bd-3c0d-4fd6-9c36-7ff1bd5c822a)
![ESERCIZI - Architettura-85](https://github.com/CS-Swap/Architettura-degli-Elaboratori/assets/110790224/3c5150d0-45f5-488a-a72e-e022e7cbe8df)
